### PR TITLE
generate fixture for PostGenerationMethodCall declaration

### LINKED ISSUE #87: TypeError: can only concatenate str (not "PostGenerationMethodCall") to str
When using factory.PostGenerationMethodCall on a factory I get the type error in the title.

The factory I'm working with is a django user model. Simple example:

```python
class UserFactory(factory.django.DjangoModelFactory):
    email = factory.Sequence(lambda n: "user-{0}@example.com".format(n))
    password = factory.PostGenerationMethodCall("set_password", ["password"])
    first_name = fuzzy.FuzzyText(length=20)
    last_name = fuzzy.FuzzyText(length=20)

    class Meta:
        model = "users.User"
        django_get_or_create = ("email",)

register(UserFactory)
```

Attempting to use that factory as a fixture will throw:
```
env/lib/python3.7/site-packages/pytest_factoryboy/plugin.py:83: in evaluate
    self.execute(request, function, deferred)
env/lib/python3.7/site-packages/pytest_factoryboy/plugin.py:65: in execute
    self.results[model][attr] = function(request)
env/lib/python3.7/site-packages/pytest_factoryboy/fixture.py:273: in deferred
    declaration.call(instance, step, context)
env/lib/python3.7/site-packages/factory/declarations.py:743: in call
    return method(*args, **kwargs)
env/lib/python3.7/site-packages/django/contrib/auth/base_user.py:98: in set_password
    self.password = make_password(raw_password)
env/lib/python3.7/site-packages/django/contrib/auth/hashers.py:78: in make_password
    return hasher.encode(password, salt)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <django.contrib.auth.hashers.MD5PasswordHasher object at 0x7f1fa3759cd0>, password = <factory.declarations.PostGenerationMethodCall object at 0x7f1fa47ff390>, salt = 'XXXXXXXXX'

    def encode(self, password, salt):
        assert password is not None
        assert salt and '$' not in salt
>       hash = hashlib.md5((salt + password).encode()).hexdigest()
E       TypeError: can only concatenate str (not "PostGenerationMethodCall") to str

env/lib/python3.7/site-packages/django/contrib/auth/hashers.py:508: TypeError
```

The error stems from https://github.com/pytest-dev/pytest-factoryboy/blob/37e46da05ca781b0034c84875c7cef08348abf58/pytest_factoryboy/fixture.py#L75 . 

Changing that check to something like:
```python
        if isinstance(value, factory.declarations.PostGeneration) or isinstance(
                value, factory.declarations.PostGenerationMethodCall
            ):
```
Seems to fix the issue. I haven't looked at this closely though so not sure if that's the best solution.

possibly relevant versions:
pytest-factoryboy==2.0.3
django==2.2.6
factory-boy==2.12.0
pytest-django==3.6.0

### PR DESCRIPTION
**Pull request by skarzi**
Closes #87 

This PR adds fixtures generation for all subclasses (except `RelatedFactory`) of `PostGenerationDeclaration`.
The fixture is generated from `attr_fixture` and it's simply passed as a `context.value` to declaration's `call()` method call, so it will be available for:

+ `PostGeneration` in `function` as `extracted`
+ `PostGenerationMethodCall`  as  first argument of called method

---
This change is [Reviewable](https://reviewable.io/reviews/pytest-dev/pytest-factoryboy/103)

### CODE REVIEW

**Comment by skarzi** [comment]
@youtux I will be very grateful if you find some time to review this PR because such a change will make working with the `PostGenerationMethodCall` declaration much easier

File: tests/test_postgen_dependencies.py [hunk 1]
```diff
@@ -76,14 +92,52 @@ def test_getfixturevalue(request, factoryboy_request):
...
 def test_after_postgeneration(foo):
     """Test _after_postgeneration is called."""
-    assert foo._postgeneration_results == {"set1": None}
+    assert foo._postgeneration_results == {"set1": None, "set2": None, "secret": None, "number": None}
     assert foo._create is True
 
 
+@pytest.mark.parametrize("foo__set2", [123])
+def test_postgeneration_fixture(foo):
+    """Test fixture for ``PosGeneration`` declaration."""
```
Comment (MRigal) [inline tests/test_postgen_dependencies.py:125]: typo PostGeneration

Comment (MRigal) [reply tests/test_postgen_dependencies.py:125]: also below

Comment (skarzi) [reply tests/test_postgen_dependencies.py]: thanks for a review ;)

**Review by olegpidsadnyi — APPROVED** [review]
looks ok to me

**Comment by skarzi** [comment]
hi :wave: 

could you prepare a new release with this feature included? 
Of course, I can help if something else should be done/implemented before the new release

**Comment by skarzi** [comment]
@youtux could you review this PR?

File: tests/test_postgen_dependencies.py [hunk 2]
```diff
@@ -107,15 +122,51 @@ def test_getfixturevalue(request, factoryboy_request: Request):
...
 
 
 def test_after_postgeneration(foo: Foo):
     """Test _after_postgeneration is called."""
     assert foo._create is True
-
-    foo._postgeneration_results["set1"] == "set to 1"
-    foo._postgeneration_results["baz"].foo is foo
-    assert len(foo._postgeneration_results) == 2
+    assert foo._postgeneration_results.pop("baz")
```
Comment (youtux) [inline tests/test_postgen_dependencies.py:150]: We used to assert that `... is foo`

File: tests/test_postgen_dependencies.py [hunk 3]
```diff
@@ -107,15 +122,51 @@ def test_getfixturevalue(request, factoryboy_request: Request):
...
+    """Test fixture for ``PostGeneration`` declaration."""
+    assert foo.value == 123
+
+
+@pytest.mark.parametrize(
+    ("foo__secret", "foo__number"),
+    [
+        ("test secret", 456),
+    ],
+)
```
Comment (youtux) [inline tests/test_postgen_dependencies.py:165]: You can make 2 different calls to parametrize. It should look more neat

File: tests/test_postgen_dependencies.py [hunk 4]
```diff
@@ -47,22 +56,29 @@ class Meta:
...
 @register
 class FooFactory(factory.Factory):
-
     """Foo factory."""
 
     class Meta:
         model = Foo
 
     value = 0
+    #: Value that is expected at the constructor
```
Comment (youtux) [inline tests/test_postgen_dependencies.py:65]: Useless comment, you can just remove it

File: pytest_factoryboy/fixture.py [hunk 5]
```diff
@@ -11,6 +11,8 @@
 import factory.declarations
 import factory.enums
 import inflection
+from factory.declarations import NotProvided
+from typing_extensions import Protocol
```
Comment (youtux) [inline pytest_factoryboy/fixture.py:15]: Unused?

File: pytest_factoryboy/fixture.py [hunk 6]
```diff
@@ -96,8 +98,22 @@ def register_(factory_class: F) -> F:
         args = []
         attr_name = SEPARATOR.join((model_name, attr))
 
-        if isinstance(value, factory.declarations.PostGeneration):
-            value = kwargs.get(attr, None)
+        if isinstance(value, factory.declarations.PostGenerationDeclaration) and not isinstance(
+            value, factory.RelatedFactory
```
Comment (youtux) [inline pytest_factoryboy/fixture.py:102]: We do another isinstqnce(..., RelatedFactory) in the else branch. Can you invert this if-else, and do this kind of check instead:

```python
if instance(..., (SubFactory, RelatedFactory)):
    ...
elif isinstance(..., PostGeneration):
    ...
elif isinstance(..., PostGenenerationMethodCall)
    ...
else:
raise NotImplementedError(useful error message to the user)
```

**Review by youtux — CHANGES_REQUESTED** [review]

**Comment by youtux** [comment]
Sorry it took so long, I didn't have it under my radar

File: tests/test_postgen_dependencies.py (hunk 4 above)
Comment (youtux) [reply tests/test_postgen_dependencies.py:65]: actually, nevermind

File: tests/test_postgen_dependencies.py (hunk 2 above)
Comment (youtux) [reply tests/test_postgen_dependencies.py]: nevermind, we didn't. I added a test for it and marked as xfail, since it's not working at the moment

**Review by youtux — APPROVED** [review]

**Comment by skarzi** [comment]
sorry for the delay from my side, but if you want I can add another PR with changes suggested by you

**Comment by youtux** [comment]
No worries, I already applied the suggestion and merged the pr. also release a new version with this fix on pypi

**Pull request merged**